src/realm/query_expression.hpp

```diff
@@ -3107,39 +3115,132 @@ class Columns<Dictionary> : public ColumnsCollection<Mixed> {
 
     void evaluate(size_t index, ValueBase& destination) override;
 
-    virtual std::string description(util::serializer::SerialisationState& state) const override
-    {
-        std::string key_str;
-        if (!m_key.is_null()) {
-            key_str = std::string(".") + std::string(m_key.get_string());
-        }
-        return ColumnListBase::description(state) + key_str;
-    }
-
     std::unique_ptr<Subexpr> clone() const override
     {
         return make_subexpr<Columns<Dictionary>>(*this);
     }
 
     Columns(Columns const& other)
         : ColumnsCollection<Mixed>(other)
-        , m_prop_list(other.m_prop_list)
-        , m_objkey(other.m_objkey)
         , m_key_type(other.m_key_type)
     {
-        if (!other.m_key.is_null()) {
-            key(other.m_key);
-        }
     }
 
 protected:
+    DataType m_key_type;
+};
+
+class ColumnDictionaryKey : public Columns<Dictionary> {
+public:
+    ColumnDictionaryKey(Mixed key_value, const Columns<Dictionary>& dict)
+        : Columns<Dictionary>(dict)
+    {
+        init_key(key_value);
+    }
+
+    ColumnDictionaryKey& property(const std::string& prop)
+    {
+        m_prop_list.push_back(prop);
+        return *this;
+    }
+
+    void evaluate(size_t index, ValueBase& destination) override;
+
+    std::string description(util::serializer::SerialisationState& state) const override
+    {
+        return ColumnListBase::description(state) + std::string(".") + std::string(m_key.get_string());
+    }
+
+    std::unique_ptr<Subexpr> clone() const override
+    {
+        return std::unique_ptr<Subexpr>(new ColumnDictionaryKey(*this));
+    }
+
+    ColumnDictionaryKey(ColumnDictionaryKey const& other)
+        : Columns<Dictionary>(other)
+        , m_prop_list(other.m_prop_list)
+        , m_objkey(other.m_objkey)
+    {
+        init_key(other.m_key);
+    }
+
+private:
     Mixed m_key;
     std::string m_buffer;
     std::vector<std::string> m_prop_list;
     ObjKey m_objkey;
-    DataType m_key_type;
+
+    void init_key(Mixed key_value);
 };
 
+// Returns the keys
+class ColumnDictionaryKeys : public Subexpr2<StringData> {
```

Comment:
Shouldn't this be a subclass of `Subexpr2<Mixed>` or how can we handle dictionaries with other key types like int?

Reply:
It should. But it turned out to be rather complicated. For the time being, we only support string keys, so no reason to make it more complicated that necessary.

Reply:
ok, I forgot we only supported string keys right now, it seems most of the implementation is ready for any type of key